### PR TITLE
Add when evaluate filter

### LINKED ISSUE #224: [Feature] Keyword / Commit message condition
This should enable a new trigger for keywords in the commit message or title (using either wildcard glob or regex). The usage is as follows:

```
trigger:
  keyword:
  - "*[RUN PIPELINE]*"
```
Should only run if the keyword is matched.

or

```
trigger:
  keyword:
    exclude:
    - "*[SKIP DEPLOY]*"
```
Should not run if the keyword is matched.

I considered simplifying the notation just matching the keyword anywhere in the string, but it would give the user less control on matching specific strings.

### LINKED ISSUE #312: `when` evaluation string filter
### Clear and concise description of the problem

Currently Woodpecker already has quite some options for `when` to filter steps, but there are already some more requested adding all kind of exceptions would require quite some maintenance in the future.

### Suggested solution

Woodpecker could introduce artithmetic/string expressions to evaluate if a step should be executed replacing the current yml options. This could be done by some library like [govaluate](https://github.com/Knetic/govaluate). By adding all kind of meta-data available for a step as variables and adding a few special functions a user could simply write own expressions like this example:

```bash
woodpecker.event == "tag" && contains(toLowerCase(woodpecker.commit.message), "[skip ci"])
```

A similar feature is available for [Github Actions](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions)

### LINKED ISSUE #963: Allow the usage of environment variables in filters
### Clear and concise description of the problem

I'm used (from `GitLab CI`) to be able to provide some some references in filters; for instance:

```yaml
default:
  image: docker.io/docker/compose:alpine-1.29.2

stages:
  - verify

validate:
  stage: verify
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
    - if: $CI_MERGE_REQUEST_ID
  script:
    - docker-compose config --quiet
```

While the filters in `Woodpecker CI` are slightly different, they achieve (almost) the same results, so it would be great to be able to use environment variables in these as well.

```yaml
pipeline:
  validate:
    commands:
      - docker-compose config --quiet
    group: verify
    image: docker.io/docker/compose:alpine-1.29.2
    when:
      branch: ${CI_REPO_DEFAULT_BRANCH}
```

### Suggested solution

Just to be able to refer to environment variables in filters — expanded by default :wink:

### Alternative

N/A

### Additional context

_No response_

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by anbraten**
closes #312 
closes #224
closes #963

Have a look for
https://github.com/antonmedv/expr/blob/master/docs/Language-Definition.md

# Changes
- [x] add docs
- [x] add samples

# Follow ups
- drop support for when.path.ignore_message

### CODE REVIEW

**Comment by anbraten** [comment]
Should we call it when.if or when.eval, ... I am open for ideas.

**Comment by 6543** [comment]
eval or expression but not if

**Comment by anbraten** [comment]
Should we call it when.if or when.eval, ... 

I am open for suggestions.

File: docs/docs/20-usage/20-pipeline-syntax.md [hunk 1]
```diff
@@ -416,6 +416,33 @@ when:
...
+
+Execute a step only if the provided evaluate expression is equal to true. Each [`CI_` variable](./50-environment.md#built-in-environment-variables) can be used inside the expression.
+
+The expression syntax can be found in [the docs](https://github.com/antonmedv/expr/blob/master/docs/Language-Definition.md) of the underlying library.
+
+Run on pushes to the default branch for the repository `owner/repo`:
+
+```yaml
+when:
+  - evaluate: 'CI_BUILD_EVENT == "push" && CI_REPO == "owner/repo" && CI_COMMIT_BRANCH == CI_REPO_DEFAULT_BRANCH'
```
Comment (6543) [inline docs/docs/20-usage/20-pipeline-syntax.md:429]:
```suggestion
  evaluate: 'CI_BUILD_EVENT == "push" && CI_REPO == "owner/repo" && CI_COMMIT_BRANCH == CI_REPO_DEFAULT_BRANCH'
```

File: docs/docs/20-usage/20-pipeline-syntax.md [hunk 2]
```diff
@@ -416,6 +416,33 @@ when:
...
+```
+
+Run on commits created by user `woodpecker-ci`:
+
+```yaml
+when:
+  - evaluate: 'CI_COMMIT_AUTHOR == "woodpecker-ci"'
```
Comment (6543) [inline docs/docs/20-usage/20-pipeline-syntax.md:436]:
```suggestion
  evaluate: 'CI_COMMIT_AUTHOR == "woodpecker-ci"'
```

File: docs/docs/20-usage/20-pipeline-syntax.md [hunk 3]
```diff
@@ -416,6 +416,33 @@ when:
...
+```
+
+Skip all commits containing `please ignore me` in the commit message:
+
+```yaml
+when:
+  - evaluate: 'not (CI_COMMIT_MESSAGE contains "please ignore me")'
```
Comment (6543) [inline docs/docs/20-usage/20-pipeline-syntax.md:443]:
```suggestion
  evaluate: 'not (CI_COMMIT_MESSAGE contains "please ignore me")'
```

Comment (6543) [reply docs/docs/20-usage/20-pipeline-syntax.md:443]: we dont need a list in examples as evaluate can express any check for itselve ...

Comment (6543) [reply docs/docs/20-usage/20-pipeline-syntax.md:443]: ... I would say?

File: docs/docs/20-usage/20-pipeline-syntax.md (hunk 1 above)
Comment (6543) [inline docs/docs/20-usage/20-pipeline-syntax.md:429]:
```suggestion
  evaluate: 'CI_BUILD_EVENT == "push" && CI_REPO == "owner/repo" && CI_COMMIT_BRANCH == CI_REPO_DEFAULT_BRANCH'
```

File: go.mod [hunk 4]
```diff
@@ -56,6 +56,7 @@ require (
 	github.com/Microsoft/go-winio v0.5.2 // indirect
 	github.com/PuerkitoBio/purell v1.1.1 // indirect
 	github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578 // indirect
+	github.com/antonmedv/expr v1.9.0 // indirect
```
Comment (6543) [inline go.mod:59]: this should be direct ... I think `go mod tidy` will fix it

File: docs/docs/20-usage/20-pipeline-syntax.md (hunk 3 above)
Comment (anbraten) [reply docs/docs/20-usage/20-pipeline-syntax.md:443]: We used the multi when format on the whole pipeline syntax page so I just used it again to be persistent.

**Review by 6543 — APPROVED** [review]

**Pull request merged**